File: partitionList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @param x, an integer
    # @return a ListNode
    def partition(self, head, x):
        less = None
        ltail = None
        geq = None
        gtail = None
        p = head
        while p:
        	q = p.next
        	p.next = None
        	if p.val < x:
        		if less is None:
        			less = p
        			ltail = p
        		else:
        			ltail.next = p
        			ltail = ltail.next
        	else:
        		if geq is None:
        			geq = p
        			gtail = p
        		else:
        			gtail.next = p
        			gtail = gtail.next
        	p = q

        if less:
        	head = less
        	ltail.next = geq
        else:
        	head = geq
        return head


h = None
tail = None
for i in range(10):
	if h is None:
		h = ListNode(i + (-2)**i)
		tail = h
	else:
		tail.next = ListNode(i + (-2)**i)
		tail = tail.next

p = h
while p:
	print p.val,
	p = p.next
print

s = Solution()
h = s.partition(h,4)
p = h
while p:
	print p.val,
	p = p.next
print

File: JAVA/partitionList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode p = head, greaterHead = null, greaterTail = null, lessHead = null, lessTail = null;
        while (p!=null){
            if (p.val >= x){
                if (greaterHead == null){
                    greaterHead = p;
                    greaterTail = p;
                }
                else{
                    greaterTail.next = p;
                    greaterTail = greaterTail.next;
                }
                p = p.next;
                greaterTail.next = null;
            }
            else{
                if (lessHead == null){
                    lessHead = p;
                    lessTail = p;
                }
                else{
                    lessTail.next = p;
                    lessTail = lessTail.next;
                }
                p = p.next;
                lessTail.next = null;
            }
        }
        if (lessHead != null){
            lessTail.next = greaterHead;
            return lessHead;
        }
        return greaterHead;
    }
}